### ISSUE TITLE
High CPU usage and low GPU usage

### ISSUE THREAD

**Issue by marinhero** [getumbrel/llama-gpt#80]
Hey team, first of all, thanks a lot for all the hard work you've put in here. This is an amazing project and makes trying new models a superb experience, thank you!

Now, I've noticed that when I run the service my CPU usage goes to 100% while my queries are being answered and GPU usage stays around 30% or 40%. I encountered the opposite while running the same questions using other tools but for some reason, llama-gpt appears to be doing all the work using my CPU. The chat experience may feel slightly slow but nothing to be a deal breaker. Albeit this CPU is beefy and I would expect the experience to degrade in other machines. **Is this kind of resource consumption expected or is this a bug?**

https://github.com/getumbrel/llama-gpt/assets/316711/8f0044f7-4d12-4b5f-9391-29d62b371480

I started the service running ` ./run.sh --model 13b --with-cuda` and whenever I stop the service via CTRL-C I can see:

```
[+] Stopping 2/2
 ✔ Container llama-gpt-llama-gpt-ui-1             Stopped                                                                                                                 0.5s
 ✔ Container llama-gpt-llama-gpt-api-cuda-ggml-1  Stopped                                                                                                                 1.3s
canceled
```

It seems to be running the CUDA-compatible images so I would expect a higher GPU usage but that doesn't appear to be the case.

Hardware/Software Info:

- GPU: NVIDIA 4090 with 24GB of Memory.
- CPU: AMD Ryzen 9 7950X3D 16-Core Processor.
- RAM: 64GB DDR5 at 4800MHz
- OS: Windows 11
- Execution Mode: Running llama-gpt via docker inside WSL2.
- Model: llama2 13B.

Let me know if you need more information, I'd be happy to provide it. Thanks!

**Comment by mayankchhabra** [comment]
Thanks for a detailed report, @marinhero! Are you able to play around with the GPU Layers value here:

https://github.com/getumbrel/llama-gpt/blob/ab4daab90870659e59423c968d96a4394f3574c9/cuda/run.sh#L46

I'd recommend increasing it by ~5 at a time. Eg. start with 15, 20, 25, 30, and so on until you see the GPU is being fully utilized and the interference speed has significantly improved. Based on your results, we can try to add some sort of logic to increase the number of offloaded GPU layers based on available VRAM.

**Comment by r0gi** [comment]
I noticed the exact same thing on a similarly powerful machine. With an RTX3080 I set `n_gpu_layers=30` on the `Code Llama 13B Chat (GGUF Q4_K_M)` model, which drastically improved inference time.  

I think a simple improvement would be to not use all cores by default, or otherwise limiting CPU usage, as all cores get maxed out during inference with the default settings.

I only wanted half of my CPU cores to be used so I set the following:
```
n_threads=$(( $(grep -c ^processor /proc/cpuinfo) / 2 ))
```
It might be worth making some smarter logic for cpu usage as well.

**Comment by marinhero** [comment]
Hey @mayankchhabra, I just performed two types of tests:

- Increase the value of n_gpu_layers 5 by 5: GPU usage went to the high 80's when I set the value to 60. Inference time improved greatly.

- Increase until performance degradation: I hit `n_gpu_layers=300` without seeing GPU utilization past 92%. Also, there's no noticeable performance increase from having this setting set to 60. 

As an idea, increasing the default value of `n_gpu_layers` according to the amount of memory of a given GPU could be a way to provide a better OOB experience. A command to get that value  is:

```
$> nvidia-smi --query-gpu=memory.total --format=csv,noheader,nounits
```

And from there it should be possible to calculate a relation between memory and gpu_layers.

Let me know if you'd like me to make a PR for it or for updating docs as @r0gi mentioned above. I'd be happy to contribute to the project.

Cheers.
